### PR TITLE
Implement patterns

### PR DESCRIPTION
**Pull request by rerobika**
Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

**Comment by yichoi** [comment]
@rerobika please rebase this PR to the master

**Comment by yichoi** [comment]
@rerobika rebase please

**Comment by clover2123** [comment]
@rerobika It may be too late, but would you fix `src/parser/esprima_cpp/esprima.cpp` according to the [esprima](https://github.com/jquery/esprima.git) (#commit_id cfd8eba831517cc89d895eb0e992ec78210a8a0d) as much as possible?
Since our parser is implemented based on the `esprima`, matching to it is highly recommended. It is also good for updating/maintenance.

**Comment by rerobika** [comment]
@clover2123 I've rebased the PR to the latest master and updated the parsing state of the object/array patterns to stay synchronized with the commit that you mentioned.

File: src/parser/ast/ArrayPatternNode.h [hunk 1]
```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2016-present Samsung Electronics Co., Ltd
```
Comment (clover2123) [inline src/parser/ast/ArrayPatternNode.h]: It's a minor fix, 2016 -> 2019

File: src/parser/ast/ObjectPatternNode.h [hunk 2]
```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2016-present Samsung Electronics Co., Ltd
```
Comment (clover2123) [inline src/parser/ast/ObjectPatternNode.h]: Same as above.

File: src/parser/ast/RegisterReferenceNode.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2016-present Samsung Electronics Co., Ltd
```
Comment (clover2123) [inline src/parser/ast/RegisterReferenceNode.h]: The last one. Other looks nice to me!

Comment (rerobika) [reply src/parser/ast/RegisterReferenceNode.h]: Thanks, I fixed it.

**Pull request merged**